### ISSUE TITLE
lora save and load

### ISSUE THREAD

**Issue by xiaobingbuhuitou** [Hoper-J/AI-Guide-and-Demos-zh_CN#1]
`
import torch
import torch.nn as nn
from peft import get_peft_model, LoraConfig, PeftModel
from accelerate import Accelerator

class Encoder(nn.Module):
    def __init__(self):
        super().__init__()
        self.fc1 = nn.Linear(10, 10)
        self.fc2 = nn.Linear(10, 10)

    def forward(self, x):
        return self.fc2(self.fc1(x))

class SimpleModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.encoder = Encoder()
        self.linear2 = nn.Linear(10, 1)

    def forward(self, x):
        x = self.encoder(x)
        return self.linear2(x)

def test_model_diff():
    # 1. 设置随机种子确保可复现
    torch.manual_seed(42)

    # 2. 初始化模型和LoRA配置
    model = SimpleModel()
    peft_config = LoraConfig(
        r=8,
        lora_alpha=32,
        target_modules=["encoder.fc1"],
        lora_dropout=0.1,
        bias="none",
        modules_to_save=["linear2"],
    )

    # 3. 创建LoRA模型
    model = get_peft_model(model, peft_config)

    # 4. 简单训练
    X = torch.randn(100, 10)
    y = torch.randn(100, 1)
    optimizer = torch.optim.Adam(model.parameters(), lr=0.01)
    criterion = nn.MSELoss()
    accelerator = Accelerator()
    model, optimizer, X, y = accelerator.prepare(model, optimizer, X, y)

    print("训练前的参数:")
    for name, param in model.named_parameters():
        if param.requires_grad:
            print(f"{name}: {param.data.mean().item():.6f}")

    # 训练
    model.train()
    X = X.to(accelerator.device)
    y = y.to(accelerator.device)
    
    for epoch in range(2):
        optimizer.zero_grad()
        output = model(X)
        loss = criterion(output, y)
        loss.backward()
        optimizer.step()
        print(f"Epoch {epoch+1}, Loss: {loss.item():.4f}")

    # 5. 保存前的输出
    test_input = torch.randn(5, 10)
    model.eval()
    test_input = test_input.to(accelerator.device)
    with torch.no_grad():
        original_output = model(test_input)
    print("\n训练后的参数:")
    for name, param in model.named_parameters():
        if param.requires_grad:
            print(f"{name}: {param.data.mean().item():.6f}")
    
    # 6. 保存模型
    save_path = "test_model"
    unwrapped_model = accelerator.unwrap_model(model)
    unwrapped_model.save_pretrained(save_path, safe_serialization=True)

    # 7. 加载模型
    new_model = SimpleModel()
    new_model = PeftModel.from_pretrained(new_model, save_path)
    new_model = accelerator.prepare(new_model)

    # 8. 验证加载后的模型
    new_model.eval()
    with torch.no_grad():
        loaded_output = new_model(test_input)

    print("\n加载后的参数:")
    for name, param in new_model.named_parameters():
        if param.requires_grad:
            print(f"{name}: {param.data.mean().item():.6f}")

    print("\n输出比较:")
    print("原始输出:", original_output)
    print("加载后输出:", loaded_output)
    print(f"最大差异: {(original_output - loaded_output).abs().max().item()}")
    print(f"平均差异: {(original_output - loaded_output).abs().mean().item()}")

    # 10. 检查模型状态
    print("\n模型模式:")
    print(f"原始模型: {model.training}")
    print(f"加载模型: {new_model.training}")
    
if __name__ == "__main__":
    test_model_diff()
`
  佬这是我的测试代码，然后使用peft==0.13.2 发现再次加载模型时缺少了lora参数，请问你有什么解决方法么？

**Comment by Hoper-J** [comment]
这段代码是成功加载了的，但存在两个误解：
1. PeftModel.save_pretrained() 并不是保存整个模型，所以如果想横向对比输出差异的话，应该先保证基础模型的参数是相同的，换句话说，就是在添加 LoRA 之前保存。
2. PeftModel.from_pretrained() 的参数 is_trainable 默认为 False，所以使用 param.requires_grad 进行判断的话将无法识别到 LoRA 层，因为此时默认不进行训练。
 
你可以非常简单的使用 print(new_model) 来查看架构。

尝试运行下面的代码：
```python
import os
import torch
import torch.nn as nn
from peft import get_peft_model, LoraConfig, PeftModel
from accelerate import Accelerator

class Encoder(nn.Module):
    def __init__(self):
        super().__init__()
        self.fc1 = nn.Linear(10, 10)
        self.fc2 = nn.Linear(10, 10)
    
    def forward(self, x):
        return self.fc2(self.fc1(x))
        
class SimpleModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.encoder = Encoder()
        self.linear2 = nn.Linear(10, 1)

    def forward(self, x):
        x = self.encoder(x)
        return self.linear2(x)

def test_model_diff():
    # 1. 设置随机种子确保可复现
    torch.manual_seed(42)
    
    # 2. 初始化基础模型并保存其初始权重
    model = SimpleModel()
    base_model_path = "base_model_initial_weights.pth"
    torch.save(model.state_dict(), base_model_path)  # 保存基础模型权重
    
    print(f"基础模型初始权重已保存至 {base_model_path}")
    
    # 3. 添加 LoRA 配置并创建 LoRA 模型
    peft_config = LoraConfig(
        r=8,
        lora_alpha=32,
        target_modules=["encoder.fc1"],
        lora_dropout=0.1,
        bias="none",
        modules_to_save=["linear2"],
    )
    model = get_peft_model(model, peft_config)
    
    # 4. 简单训练
    X = torch.randn(100, 10)
    y = torch.randn(100, 1)
    optimizer = torch.optim.Adam(model.parameters(), lr=0.01)
    criterion = nn.MSELoss()
    accelerator = Accelerator()
    model, optimizer, X, y = accelerator.prepare(model, optimizer, X, y)
    
    print("\n训练前的参数:")
    for name, param in model.named_parameters():
        if param.requires_grad:
            print(f"{name}: {param.data.mean().item():.6f}")
    
    # 训练
    model.train()
    X = X.to(accelerator.device)
    y = y.to(accelerator.device)
    
    for epoch in range(2):
        optimizer.zero_grad()
        output = model(X)
        loss = criterion(output, y)
        loss.backward()
        optimizer.step()
        print(f"Epoch {epoch+1}, Loss: {loss.item():.4f}")
    
    # 5. 保存前的测试输出
    test_input = torch.randn(5, 10).to(accelerator.device)
    model.eval()
    with torch.no_grad():
        original_output = model(test_input)
    print("\n训练后的参数:")
    for name, param in model.named_parameters():
        if param.requires_grad:
            print(f"{name}: {param.data.mean().item():.6f}")
    
    # 6. 保存训练后的 LoRA 模型
    save_path = "trained_model_with_lora"
    accelerator.wait_for_everyone()
    unwrapped_model = accelerator.unwrap_model(model)
    unwrapped_model.save_pretrained(save_path, safe_serialization=True)
    torch.save(unwrapped_model.state_dict(), os.path.join(save_path, 'pytorch_model.bin'))
    
    # 7. 加载基础模型初始权重并添加 LoRA 参数
    new_model = SimpleModel()
    new_model.load_state_dict(torch.load(base_model_path))  # 加载基础模型初始权重
    print(f"\n基础模型初始权重已加载")
    print("\n打印加载 LoRA 前的模型架构:")
    print(new_model)
    
    new_model = PeftModel.from_pretrained(new_model, save_path)  # 加载 LoRA 参数
    new_model = accelerator.prepare(new_model)
    print("\n打印加载 LoRA 后的模型架构:")
    print(new_model)
    
    # 8. 验证加载后的模型
    new_model.eval()
    with torch.no_grad():
        loaded_output = new_model(test_input)
    
    print("\n加载后的参数:")
    for name, param in new_model.named_parameters():
        if param.requires_grad:
            print(f"{name}: {param.data.mean().item():.6f}")
    
    print("\n输出比较:")
    print("原始输出:", original_output)
    print("加载后输出:", loaded_output)
    print(f"最大差异: {(original_output - loaded_output).abs().max().item()}")
    print(f"平均差异: {(original_output - loaded_output).abs().mean().item()}")

if __name__ == "__main__":
    test_model_diff()
```

输出：
```
基础模型初始权重已保存至 base_model_initial_weights.pth

训练前的参数:
base_model.model.encoder.fc1.lora_A.default.weight: -0.008933
base_model.model.encoder.fc1.lora_B.default.weight: 0.000000
base_model.model.linear2.modules_to_save.default.weight: -0.046176
base_model.model.linear2.modules_to_save.default.bias: -0.256834
Epoch 1, Loss: 1.0215
Epoch 2, Loss: 0.9895

训练后的参数:
base_model.model.encoder.fc1.lora_A.default.weight: -0.008933
base_model.model.encoder.fc1.lora_B.default.weight: 0.000245
base_model.model.linear2.modules_to_save.default.weight: -0.046320
base_model.model.linear2.modules_to_save.default.bias: -0.237050

基础模型初始权重已加载

打印加载 LoRA 前的模型架构:
SimpleModel(
  (encoder): Encoder(
    (fc1): Linear(in_features=10, out_features=10, bias=True)
    (fc2): Linear(in_features=10, out_features=10, bias=True)
  )
  (linear2): Linear(in_features=10, out_features=1, bias=True)
)

打印加载 LoRA 后的模型架构:
PeftModel(
  (base_model): LoraModel(
    (model): SimpleModel(
      (encoder): Encoder(
        (fc1): lora.Linear(
          (base_layer): Linear(in_features=10, out_features=10, bias=True)
          (lora_dropout): ModuleDict(
            (default): Dropout(p=0.1, inplace=False)
          )
          (lora_A): ModuleDict(
            (default): Linear(in_features=10, out_features=8, bias=False)
          )
          (lora_B): ModuleDict(
            (default): Linear(in_features=8, out_features=10, bias=False)
          )
          (lora_embedding_A): ParameterDict()
          (lora_embedding_B): ParameterDict()
          (lora_magnitude_vector): ModuleDict()
        )
        (fc2): Linear(in_features=10, out_features=10, bias=True)
      )
      (linear2): ModulesToSaveWrapper(
        (original_module): Linear(in_features=10, out_features=1, bias=True)
        (modules_to_save): ModuleDict(
          (default): Linear(in_features=10, out_features=1, bias=True)
        )
      )
    )
  )
)

加载后的参数:
base_model.model.linear2.modules_to_save.default.weight: -0.046320
base_model.model.linear2.modules_to_save.default.bias: -0.237050

输出比较:
原始输出: tensor([[ 0.0193],
        [-0.0488],
        [-0.3356],
        [ 0.0752],
        [-0.1869]], device='cuda:0')
加载后输出: tensor([[ 0.0193],
        [-0.0488],
        [-0.3356],
        [ 0.0752],
        [-0.1869]], device='cuda:0')
最大差异: 0.0
平均差异: 0.0
```

**Comment by xiaobingbuhuitou** [comment]
感谢佬的回答，确实是这个问题。感谢
